### PR TITLE
Add analytic pulse function samplers

### LINKED ISSUE #1935: Standard sampling library
### What is the expected enhancement?
A library of standard `Sampler` (#1934) strategies should be provided for users that transform a continuous function of `t` into a `FunctionPulse`.

Pulses:
- [x] `left`
- [x] `right`
- [x] `mid_point`

### PR DESCRIPTION
**Pull request by taalexander**
### Summary
Add analytic pulse sampler which convert an analytic pulse of signature:
`def f(times: np.ndarray, *args, **kwargs) -> np.ndarray`
to a discrete pulse:
`def g(duration: int, *args, **kwargs) -> SamplePulse`

These will be used to implement the discrete pulse library from analytic pulse functions. 

Closes #1935.
### Details and comments
Currently implemented samplers:
- `left`
- `right`
- `midpoint`

These are decorators that may be used like:
```python
@pulse.samplers.left
def linear(x: np.ndarray, m: float, b: float) -> np.ndarray:
    """Linear test function
    Args:
        x: Input times.
        m: Slope.
        b: Intercept
    Returns:
        np.ndarray
    """
    return m*x+b
```

### CODE REVIEW

**Comment by nkanazawa1989** [comment]
Can you add `name` to `sampler`? just to have the same interface with `functional_pulse`.

**Comment by taalexander** [comment]
Since the `sampler` just returns a `functional_pulse` it already has the same interface as `functional_pulse`.

File: qiskit/pulse/samplers/decorators.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+def sampler(sample_function: Callable) -> Callable:
+    """Sampler decorator base method.
+
+    Samplers are used for converting an continuous function to a discretized pulse.
+
+    They operate on a function with the signature:
+        `def f(times: np.ndarray, *args, **kwargs) -> np.ndarray`
+    Where `times` is a numpy array of floats with length `n_times` and the output array
+    is a complex numpy array with length `n_times`. The output of the decorator is an
```
Comment (ajavadia) [inline qiskit/pulse/samplers/decorators.py]: update documentation (`n_times` not defined anywhere)

File: qiskit/pulse/samplers/decorators.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        # but still get rest of benefits of wraps
+        # such as __name__, __qualname__
+        call_sampler.__dict__.pop('__wrapped__')
+        # wrap with functional pulse
+        return commands.functional_pulse(call_sampler)
+
+    return generate_sampler
+
+
+def left_sample_fun(continuous_pulse: Callable, duration: int, *args, **kwargs) -> np.ndarray:
```
Comment (ajavadia) [inline qiskit/pulse/samplers/decorators.py]: seems like this function should be deleted.

**Comment by ajavadia** [comment]
can you help me understand this:
        ```left_linear_pulse_fun = samplers.left(linear)
        pulse = left_linear_pulse_fun(duration, m=m, b=b)```

So `linear` is just a function (not a pulse yet). Why does it need to have the `times` argument, which is characteristic of a discrete pulse?

And then in the above, do we get an instance of a continuous pulse somewhere?

I may be a bit lost here but what benefit do decorators have over just something simple like pulse objects having methods
```
a_sample_pulse = a_continuous_pulse.sample(strategy='left')
```

**Comment by taalexander** [comment]
Times is characteristic of a continuous pulse and is an array of times to evaluate the continuous function with.  Discrete pulses accept an integer `duration`. By `sampling` we now have an object that may be evaluated to return decorators.

The approach you suggest means having another type of `Instruction` class to handle. It makes plotting harder as well. I would like to go in this direction, in the long run, to support analytic pulses which have a duration and an underlying set of parameters (same parameters as parameterized circuits) that may then be sampled in the way you suggest to output a `SamplePulse`, or passed as an analytic pulse directly if supported by the backend.

**Comment by ajavadia** [comment]
Ok i'm fine moving forward. Tagged @kdk in case he has comments on the design, but I think we can also revisit later.

**Pull request merged**